lib/UART/src/UART.cpp

```diff
@@ -0,0 +1,217 @@
+    tx_index = 0;
+}
+
+void service_transmit_UART(void)
+{
+    if (tx_string == nullptr)
+    {
+        return;
+    }
+
+    if (queue_is_full(&tx_queue))
+    {
+        //ERROR: Buffer overflow
+        return;
+    }
+    else
+    {
+        UART_print_char(tx_string[tx_index++]);
+
+        if(tx_string[tx_index] == '\0')
+        {
+            tx_index = 0;
+            tx_string = nullptr;
+            return;
+        }
+    }
+}
+
+ISR(USART_UDRE_vect)
+{  
+    dequeue_return_t tx_letter = dequeue(&tx_queue);
+    UDR0 = tx_letter.value;
+
+    if(queue_is_empty(&tx_queue))
+    {
+        UART_desable_transmit_interrupt();
+    }
+}
+
+static void reverse_string(char *str) 
+{
+    int length = 0;
+    while (str[length] != '\0')
+    {
+        length++;
+    }
+
+    int start = 0;
+    int end = length -1;
+    while(start < end)
+    {
+        char temp = str[start];
+        str[start] = str[end];
+        str[end] = temp;
+        
+        start++;
+        end--;
+    }
+}
+
+static void int_to_string(int16_t num, char *str)
+{
+    int i = 0;
+    bool is_negative = false;
+
+    if (num < 0)
+    {
+        is_negative = true;
+        num = -num;
+    }
+
+    do
+    {
+        str[i++] = num % 10 + '0';
+        num = num / 10;
+    } while (num != 0);
+
+    if (is_negative)
+    {
+        str[i++] = '-';
+    }
+    str[i] = '\0';
+    reverse_string(str);
+}
+
+void UART_printf(const char* format, ...)
+{
+    va_list args;
+    va_start(args, format);
+
+    while (*format != '\0')
+    {
+        if (*format == '%')
+        {
+            format++;
+            switch (*format)
+            {
+                case 'd': {
+                    uint16_t num = va_arg(args, uint16_t);
+                    char num_str[12];
+                    int_to_string(num, num_str);
+                    UART_print_string(num_str);
+                    break;
+                }
+                case 'f': {
+                    //double num = va_arg(args, double);
+                    //TODO
+                    break;
+                }
+                case 's': {
+                    const char *str = va_arg(args, const char *);
+                    UART_print_string(str);
+                    break;
+                }
+                default:
+                    UART_print_string("Unsupported format specifier");
+                    break;
+            }
+        } 
+        else
+        {
+            UART_print_char(*format);
+        }
+        format++;
+    }
+    va_end(args);
+}
+
+
+uint8_queue_t rx_queue;
+ISR(USART_RX_vect)
+{
+    char receivedChar = UDR0;
+    add_to_queue(&rx_queue, (uint8_t)receivedChar);
+}
+
+#define RX_BUFFER_SIZE 64
+static char rx_buffer[RX_BUFFER_SIZE];
+static uint8_t rx_index = 0;
+
+void service_receive_UART(void)
+{
+    if (queue_is_empty(&rx_queue))
+    {
+        return;
+    }
+
+    dequeue_return_t rx_letter = dequeue(&rx_queue);
+    char character = rx_letter.value;
```

Comment:
* A variable with a small scope should have a short name since it is very easy to find the source. Inversely, a variable with a large scope needs to have a very descriptive name.
* Similarly: A function with a small scope should have a more descriptive name since it's more specific and explainable. A function with a large scope should have a shorter name since it's more likely used in a lot of places, and known by the caller.

With that said, I would name these

```suggestion
    dequeue_return_t letter = dequeue(&rx_queue);
    char c = letter.value;
```